### ISSUE TITLE
Swapping Dialog elements results in second Dialog not appearing (Vue)

### ISSUE THREAD

**Issue by r-thomson** [tailwindlabs/headlessui#2933]
**What package within Headless UI are you using?**

@headlessui/vue

**What version of that package are you using?**

1.7.17

**What browser are you using?**

Firefox 122, Safari 17.2

**Reproduction URL**

https://codesandbox.io/p/devbox/weathered-forest-v3lvy2

**Describe your issue**

I am using a pattern where modals are opened and closed with `v-if` (for reasons why, see https://github.com/tailwindlabs/headlessui/discussions/2027).

```vue
<Dialog v-if="state === 1" open />
<Dialog v-if="state === 2" open />
```

One thing that I'm doing in several places is closing one modal while simultaneously opening another one (e.g. `state++`). As of v1.7.17, when trying to do that the second modal doesn't appear. See the linked reproduction for an example.

I can confirm that this behavior is not present in v1.7.16.

**Comment by SpiritusDeos** [comment]
Still have the issue in the new minor versions, is it planned to be fix soon ?

**Comment by JapSeyz** [comment]
I am experiencing this issue as well. It is seemingly coming from https://github.com/tailwindlabs/headlessui/pull/2700

**Comment by thecrypticace** [comment]
Sorry about that! One of our changes cause a timing issue when switching dialogs in the same "tick". Fixed in @headlessui/vue v1.7.21

```
npm install @headlessui/vue@latest
```
